### ISSUE TITLE
[core] Error reporting implementation

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#3923]
Implementation notes for RFC #3816 

Nice to haves in the CLI:
- [x] Display current log level
- [ ] Display PMD version
- [ ] Log command-line arguments
- [ ] Log classpath (also, avoid printing entire classpath in 'cannot resolve resource' error)
- [ ] Colors for errors when in a terminal, also a way to disable them
- [x] #3828 

Implementation
- [ ] Use MessageReporter in more places, eg RuleApplicator
- [ ] Make the MessageReporter configurable (done in #3892)
- [ ] Make a MessageReporter implementation that doesn't depend on SLF4J, use that for the CLI
